main/lsp/LSPInput.h

```diff
@@ -0,0 +1,81 @@
+#ifndef RUBY_TYPER_LSP_LSPINPUT_H
+#define RUBY_TYPER_LSP_LSPINPUT_H
+
+#include "absl/synchronization/mutex.h"
+#include <deque>
+#include <memory>
+#include <vector>
+
+namespace spdlog {
+class logger;
+};
+
+namespace sorbet::realmain::lsp {
+class LSPMessage;
+class ResponseMessage;
+class NotificationMessage;
+
+/**
+ * Interface for receiving messages from the client. It is _not_ thread safe, and assumes that only one thread will be
+ * reading from it at a time.
+ */
+class LSPInput {
+public:
+    LSPInput() = default;
+    virtual ~LSPInput() = default;
+    /**
+     * Blockingly reads the next message from the client with the given timeout (in milliseconds). Returns nullptr if no
+     * message read within timeout.
+     *
+     * Throws a FileReadException on error or EOF.
+     */
+    virtual std::unique_ptr<LSPMessage> read(int timeoutMs = 100) = 0;
+};
+
+/**
+ * Reads messages from a file descriptor (like stdin).
+ *
+ * Throws a FileReadException on error or EOF.
+ */
+class LSPFDInput final : public LSPInput {
+    // Contains unparsed strings containing a partial message read from file descriptor.
+    std::string buffer;
+    // Used to log debug messages.
+    const std::shared_ptr<spdlog::logger> logger;
+
+public:
+    const int inputFd;
+    LSPFDInput(std::shared_ptr<spdlog::logger> logger, int inputFd);
+
+    std::unique_ptr<LSPMessage> read(int timeoutMs) override;
+};
+
+/**
+ * Input is provided programmatically via the `write` method. Threadsafe.
+ *
+ * Throws a FileReadException when stream has been closed.
+ */
+class LSPProgrammaticInput final : public LSPInput {
```

Comment:
This class is unused so far. I plan to use it in LSPWrapper. Can delurk if it adds too much noise to PR.